units/src/amount/mod.rs

```diff
@@ -38,6 +38,48 @@ pub use self::{
     unsigned::Amount,
 };
 
+/// Result of an operation on [`Amount`] or [`SignedAmount`].
+///
+/// The type parameter `T` should be normally `Amout` or `SignedAmount`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AmountOpResult<T> {
+    /// Result of a successful mathematical operation.
+    Valid(T),
+    /// Result of an unsuccessful mathematical operation.
+    Error(AmountOpError),
+}
+
+impl<T> AmountOpResult<T> {
+    /// Returns the contained valid amount.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result is an `Error`.
+    pub fn unwrap_valid(self) -> T {
+        use AmountOpResult::*;
+
+        match self {
+            Valid(amount) => amount,
+            Error(_) => panic!("TODO"),
+        }
+    }
+
+    /// Converts this `AmountOpResult` to a `Result<T, AmountOpError>`.
+    pub fn valid(self) -> Result<T, AmountOpError> {
+        use AmountOpResult::*;
+
+        match self {
+            Valid(amount) => Ok(amount),
+            Error(e) => Err(e)
+        }
+    }
+}
+
+/// An error occurred while doing a mathematical operation.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct AmountOpError;
+
```

Comment:
You're missing `Sum<Amount> for AmountOpResult<Amount>`, `Sum for AmountOpResult`, `Sub<Amount> for AmountOpResult<Amount>`, `Sub for AmountOpResult`, `Mul<u64> for AmountOpResult<Amount>`, `Div<u64> for AmountOpResult`, `Rem<u64> for AmountOpResult`, and the same for `SignedAmount` as well as `Neg for AmountOpResult<SignedAmount>`.

These were the entire point of the type: so longer expressions like `alice_output.value + bob_output.value + carol_output.value` are convenient since you only check the result at the end (worsening debugging but I have ideas about that part, so don't worry).

Reply:
I believe we've got all these covered now.